### PR TITLE
Added solution

### PR DESCRIPTION
**Pull request by slaybrute**

### CODE REVIEW

File: src/main/java/core/basesyntax/dao/FruitDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+public class FruitDaoImpl implements FruitDao {
+    @Override
+    public void add(String name, int amount) {
+        Integer amountInStorage = Storage.fruits.get(name);
+        if (amountInStorage != null) {
+            Storage.fruits.put(name, amountInStorage + amount);
+        } else {
+            Storage.fruits.put(name, amount);
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/dao/FruitDaoImpl.java:13]:
```suggestion
        int currentAmount = Storage.fruits.getOrDefault(name, 0);
        Storage.fruits.put(name, currentAmount + amount);
```

File: src/main/java/core/basesyntax/db/Storage.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.db;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Storage {
+    public static final Map<String, Integer> fruits = new HashMap<>();
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/Storage.java:7]: if you have dao, make this private

File: src/main/java/core/basesyntax/reportcreator/ReportCreatorImpl.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        StringBuilder stringBuilder = new StringBuilder("fruit,quantity");
+        for (String key : Storage.fruits.keySet()) {
+            stringBuilder.append("\n")
+                    .append(key)
+                    .append(SEPARATOR).append(Storage.fruits.get(key));
+        }
+        return stringBuilder.toString();
+    }
+}
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/reportcreator/ReportCreatorImpl.java:19]:
```suggestion
```

File: src/main/java/core/basesyntax/reportcreator/ReportCreatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import core.basesyntax.db.Storage;
+
+public class ReportCreatorImpl implements ReportCreator {
+    private static final String SEPARATOR = ",";
+
+    @Override
+    public String generateReport() {
+        StringBuilder stringBuilder = new StringBuilder("fruit,quantity");
+        for (String key : Storage.fruits.keySet()) {
+            stringBuilder.append("\n")
```
Comment (okuzan) [inline src/main/java/core/basesyntax/reportcreator/ReportCreatorImpl.java:12]: Use System.lineSeparator()

File: src/main/java/core/basesyntax/reportcreator/ReportCreatorImpl.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package core.basesyntax.reportcreator;
+
+import core.basesyntax.db.Storage;
+
+public class ReportCreatorImpl implements ReportCreator {
+    private static final String SEPARATOR = ",";
+
+    @Override
+    public String generateReport() {
+        StringBuilder stringBuilder = new StringBuilder("fruit,quantity");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/reportcreator/ReportCreatorImpl.java:10]: all literals should be constant, make it a HEADER

File: src/main/java/core/basesyntax/workwithfile/filereader/CsvFileReader.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CsvFileReader implements FileReader {
+    private static final String SEPARATOR = ",";
+    private static final int INDEX_OF_HEADING = 0;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/workwithfile/filereader/CsvFileReader.java:11]:
```suggestion
    private static final int HEADER_INDEX = 0;

```

File: src/main/java/core/basesyntax/reportcreator/ReportCreator.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.reportcreator;
+
+public interface ReportCreator {
+    public String generateReport();
```
Comment (okuzan) [inline src/main/java/core/basesyntax/reportcreator/ReportCreator.java:4]: if you have generate as method name, call it ReportGenerator

File: src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
+package core.basesyntax.fruittransact;
+
+import java.util.List;
+
+public class FruitTransactionService {
+    private static final int INDEX_OF_TYPE = 0;
+    private static final int INDEX_OF_FRUIT = 1;
+    private static final int INDEX_OF_AMOUNT = 2;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java:8]: you can avoid OF by simply changing the word order
```suggestion
    private static final int TYPE_INDEX = 0;
    ...
```

File: src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                purchaseTransaction.transact(name, amount);
+                break;
+            case "s":
+            case "r":
+                supplyTransaction.transact(name, amount);
+                break;
+            default:
+                throw new RuntimeException("No such transaction type: " + type);
+
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java:28]: that's not how Strategy pattern can be implemented. You need to have a correspondence table - which handler handles each type of transaction. It will be chosen dynamically depending on the type of each transaction

File: src/main/java/core/basesyntax/fruittransact/FruitTransaction.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.fruittransact;
+
+public interface FruitTransaction {
+    void transact(String name, int amount);
+}
```
Comment (okuzan) [inline src/main/java/core/basesyntax/fruittransact/FruitTransaction.java:5]: FruitTransaction is a DTO-like model, not an interface. This model has three parameters - fruit, amount, type. Your best bet would be to follow an implementation outlined in the project description, however, you are free to do it your own way, but your current implementation can be confusing.
What you are trying to achieve here is usually called a handler (here - `OperationHandler`)- with the method `handle`.

**Review by okuzan — CHANGES_REQUESTED** [review]
Divide reading login into two parts - FileReader is concerned with just reading raw content as is. And TransactionParser forms a collection of model objects - transactions. Create an enum for the operation type. Instead of a switch case, use a hashmap to keep correspondence between the operation type and the appropriate handler for it. 
Put resources in src/main/resources folder

File: input1.csv
Comment (olekskov) [inline input1.csv:1]: csv files should be in resources folder, not in the root

File: src/main/java/core/basesyntax/workwithfile/filewriter/CsvFileWriter.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package core.basesyntax.workwithfile.filewriter;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+
+public class CsvFileWriter implements FileWriter {
```
Comment (olekskov) [inline src/main/java/core/basesyntax/workwithfile/filewriter/CsvFileWriter.java:7]: Code of this class has nothing specific for csv, it will work fine with txt, xml or anything else
```suggestion
public class FileWriterImpl implements FileWriter {

```

File: src/main/java/core/basesyntax/fruittransact/FruitTransaction.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.fruittransact;
+
+public interface FruitTransaction {
+    void handle(String name, int amount);
+}
```
Comment (olekskov) [inline src/main/java/core/basesyntax/fruittransact/FruitTransaction.java:5]: transaction should be model, not inteface
it should hold type of operation, name and amount of fruit

File: src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+                break;
+            case "s":
+            case "r":
+                fruitTransaction = new SupplyTransaction();
+                break;
+            default:
+                throw new RuntimeException("No such transaction type: " + type);
+        }
+        fruitTransaction.handle(name, amount);
+    }
```
Comment (olekskov) [inline src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java:27]: here strategy pattern should be used instead of switch/case

**Review by olekskov — CHANGES_REQUESTED** [review]
Strategy pattern usage is requirement of this task.
Please feel free to ask in chat or on Q&A session if you have any problems with it

File: src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java (hunk 13 above)
Comment (slaybrute) [reply src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java:27]: > here strategy pattern should be used instead of switch/case

I've already understood what you meant

File: src/main/java/core/basesyntax/fruittransact/strategy/FruitTransactionStrategyImpl.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
+package core.basesyntax.fruittransact.strategy;
+
+import core.basesyntax.fruittransact.FruitService;
+import core.basesyntax.fruittransact.PurchaseService;
+import core.basesyntax.fruittransact.RemnantService;
+import core.basesyntax.fruittransact.SupplyService;
+
+public class FruitTransactionStrategyImpl implements FruitTransactionStrategy {
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/fruittransact/strategy/FruitTransactionStrategyImpl.java:8]: The strategy should look like this:

```
public interface OperationHandler {
    void handle(FruitTransaction transaction);
}

public interface OperationStrategy {
    OperationHandler get(FruitTransaction.Operation operation);
}

public class OperationStrategyImpl implements OperationStrategy {
    private Map<FruitTransaction.Operation, OperationHandler> operationHandlerMap;

    public OperationStrategyImpl(Map<FruitTransaction.Operation,
            OperationHandler> operationHandlerMap) {
        this.operationHandlerMap = operationHandlerMap;
    }

    @Override
    public OperationHandler get(FruitTransaction.Operation operation) {
        return operationHandlerMap.get(operation);
    }
}
```

File: src/main/java/core/basesyntax/fruittransact/FruitTransaction.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    public FruitTransaction(FruitTransactionStrategy fruitTransactionStrategy) {
+        this.fruitTransactionStrategy = fruitTransactionStrategy;
+    }
+
+    public void handle(String type, String name, int amount) {
+        FruitService fruitService = fruitTransactionStrategy.get(type);
+        fruitService.handle(name, amount);
+    }
+
+    public void handleAll(List<String[]> data) {
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/fruittransact/FruitTransaction.java:21]: It is better to use the model(`FruitTransaction`) than an array of Strings

```
public class FruitTransaction {
    private Operation operation;
    private String fruit;
    private int quantity;

    public Operation getOperation() {
        return operation;
    }

    public void setOperation(Operation operation) {
        this.operation = operation;
    }

    public String getFruit() {
        return fruit;
    }

    public void setFruit(String fruit) {
        this.fruit = fruit;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }

    public enum Operation {
        BALANCE("b"),
        SUPPLY("s"),
        PURCHASE("p"),
        RETURN("r");

        private final String operation;

        Operation(String operation) {
            this.operation = operation;
        }

        public String getOperation() {
            return operation;
        }

        public static Operation getByCode(String character) {
            for (Operation operation : values()) {
                if (operation.getOperation().equals(character)) {
                    return operation;
                }
            }
            throw new RuntimeException("Unknown operation");
        }
    }
}
```

File: src/main/java/core/basesyntax/fruittransact/PurchaseService.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+import core.basesyntax.dao.FruitDao;
+import core.basesyntax.dao.FruitDaoImpl;
+
+public class PurchaseService implements FruitService {
+    private FruitDao fruitDao = new FruitDaoImpl();
+
+    @Override
+    public void handle(String name, int amount) {
+        fruitDao.remove(name, amount);
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/fruittransact/PurchaseService.java:11]: If you have enough fruits, you can buy them
otherwise, you should throw the "Not enough fruitName to purchase" exception

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
The project should have packages (service, db, model, strategy...)
package service should have: FruitShopService, ParserService, ReaderService, ReportService, WriterService
package db should have: Storage
package model should have: FruitTransaction
package strategy should have: OperationStrategy , OperationHandler
all implementation should be to move to package impl
this will be a good code decomposition.

File: input1.csv [hunk 17]
```diff
@@ -0,0 +1,9 @@
+type,fruit,quantity
```
Comment (okuzan) [inline input1.csv:1]: not fixed, these three files should be in src/main/resources

File: src/main/java/core/basesyntax/fruittransact/strategy/FruitTransactionStrategyImpl.java (hunk 14 above)
Comment (okuzan) [reply src/main/java/core/basesyntax/fruittransact/strategy/FruitTransactionStrategyImpl.java:8]: use this suggested implementation instead of a switch case

File: src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java (hunk 13 above)
Comment (okuzan) [reply src/main/java/core/basesyntax/fruittransact/FruitTransactionService.java]: not fixed

File: src/main/java/core/basesyntax/service/SupplyService.java [hunk 18]
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dao.FruitDao;
+import core.basesyntax.dao.FruitDaoImpl;
+
+public class SupplyService implements OperationService {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/SupplyService.java:6]:
```suggestion
public class SupplyOperationHandler implements OperationHandler {
```
change everywhere

File: src/main/java/core/basesyntax/service/FileReaderImpl.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
+package core.basesyntax.service;
+
+import core.basesyntax.model.FruitTransaction;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+
+public class FileReaderImpl implements FileReader {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/FileReaderImpl.java:10]: Not fixed, Please divide the logic here - reader should read whatever is provided as is, we should separate our concerns. Move parsing data to transactions to a separate service

File: src/main/java/core/basesyntax/fruittransact/strategy/FruitTransactionStrategy.java [hunk 20]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.fruittransact.strategy;
+
+import core.basesyntax.fruittransact.FruitService;
+
+public interface FruitTransactionStrategy {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/fruittransact/strategy/FruitTransactionStrategy.java:5]: remove this and implementation, you have another strategy now

File: src/main/java/core/basesyntax/fruittransact/FruitTransaction.java [hunk 21]
```diff
@@ -0,0 +1,28 @@
+package core.basesyntax.fruittransact;
+
+import core.basesyntax.fruittransact.strategy.FruitTransactionStrategy;
+import java.util.List;
+
+public class FruitTransaction {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/fruittransact/FruitTransaction.java:6]: remove this class, you have another FruitTransaction

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/model/FruitTransaction.java [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+
+        Operation(String operation) {
+            this.operation = operation;
+        }
+
+        public String getOperation() {
+            return operation;
+        }
+
+        public static Operation getByCode(String character) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/model/FruitTransaction.java:54]:
```suggestion
        public static Operation getByCode(String code) {
```

File: src/main/java/core/basesyntax/model/FruitTransaction.java [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+            for (Operation operation : values()) {
+                if (operation.getOperation().equals(character)) {
+                    return operation;
+                }
+            }
+            throw new RuntimeException("Unknown operation");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/model/FruitTransaction.java:60]:
```suggestion
            throw new NoSuchElementException("Unknown operation: "+ code);
```

File: src/main/java/core/basesyntax/db/Storage.java [hunk 24]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Storage {
+    private static final Map<String, Integer> fruits = new HashMap<>();
+
+    public static Map<String, Integer> getFruits() {
+        return fruits;
+    }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/db/Storage.java:11]: this defeats the purpose of having a FruitDao, if you can call the map directly. Try to remove this getter, and for all the places that used it, add a required functionality to Dao

File: src/main/java/core/basesyntax/service/FileWriterImpl.java [hunk 25]
```diff
@@ -0,0 +1,16 @@
+package core.basesyntax.service;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+
+public class FileWriterImpl implements FileWriter {
+    @Override
+    public void writeToFile(File file, String report) {
+        try (BufferedWriter bufferedWriter = new BufferedWriter(new java.io.FileWriter(file))) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/FileWriterImpl.java:10]:
```suggestion
        try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(file))) {
```
import it

File: src/main/java/core/basesyntax/service/ParseService.java [hunk 26]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax.service;
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.List;
+
+public interface ParseService {
+    List<FruitTransaction> parse(List<String> fruits);
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/ParseService.java:8]:
```suggestion
```

File: src/main/java/core/basesyntax/service/ParseServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,26 @@
+package core.basesyntax.service;
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ParseServiceImpl implements ParseService {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/ParseServiceImpl.java:7]: Well done!

File: src/main/java/core/basesyntax/service/ParseServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+    @Override
+    public List<FruitTransaction> parse(List<String> fruits) {
+        ArrayList<FruitTransaction> fruitTransactions = new ArrayList<>();
+        fruits.remove(HEADER_INDEX);
+        for (String fruit : fruits) {
+            String[] parsedFruit = fruit.split(SEPARATOR);
+            fruitTransactions.add(new FruitTransaction(parsedFruit[FRUIT_INDEX],
+                    Integer.parseInt(parsedFruit[AMOUNT_INDEX]),
+                    FruitTransaction.Operation.getByCode(parsedFruit[OPERATION_INDEX])));
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/ParseServiceImpl.java:22]: integer.parseint throws numberformatexception, your getByCode - NoSuchElementException. You will have to write tests in the next HW for invalid input and edge cases, so think about how do you handle it here

File: src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+                return new RemnantOperationHandler();
+            case SUPPLY:
+            case RETURN:
+                return new SupplyOperationHandler();
+            case PURCHASE:
+                return new PurchaseOperationHandler();
+            default:
+                throw new RuntimeException("No such operation: " + operation);
+        }
+    }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/strategy/OperationStrategyImpl.java:25]: Not fixed, use a hashmap, where Operation is a key, Handler is a value

**Review by okuzan — CHANGES_REQUESTED** [review]
You're on the right track! Good job on your latest commit. Some further improvements are suggested 🧗‍♂️

File: src/main/java/core/basesyntax/service/FileReaderImpl.java [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+    private static final String SEPARATOR = ",";
+    private static final int HEADER_INDEX = 0;
+    private static final int OPERATION_INDEX = 0;
+    private static final int FRUIT_INDEX = 1;
+    private static final int AMOUNT_INDEX = 2;
+
+    @Override
+    public List<String> readFromFile(File file) {
+        ArrayList<FruitTransaction> fruitTransactions = new ArrayList<>();
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/service/FileReaderImpl.java:19]: useless value
```suggestion
```

File: src/main/java/core/basesyntax/service/PurchaseOperationHandler.java [hunk 31]
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dao.FruitDao;
+import core.basesyntax.dao.FruitDaoImpl;
+
+public class PurchaseOperationHandler implements OperationHandler {
+    private FruitDao fruitDao = new FruitDaoImpl();
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/service/PurchaseOperationHandler.java:7]: The code should not be implementation-dependent 
use a constructor for initialization
```suggestion
    private FruitDao fruitDao;
    
    public PurchaseOperationHandler(FruitDao fruitDao) {
          this.fruitDao = fruitDao;
    }
```

File: src/main/java/core/basesyntax/service/RemnantOperationHandler.java [hunk 32]
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dao.FruitDao;
+import core.basesyntax.dao.FruitDaoImpl;
+
+public class RemnantOperationHandler implements OperationHandler {
+    private FruitDao fruitDao = new FruitDaoImpl();
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/service/RemnantOperationHandler.java:7]: the same

File: src/main/java/core/basesyntax/service/SupplyOperationHandler.java [hunk 33]
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dao.FruitDao;
+import core.basesyntax.dao.FruitDaoImpl;
+
+public class SupplyOperationHandler implements OperationHandler {
+    private FruitDao fruitDao = new FruitDaoImpl();
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/service/SupplyOperationHandler.java:7]: also

File: src/main/java/core/basesyntax/service/ReportServiceImpl.java [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+
+import core.basesyntax.dao.FruitDaoImpl;
+
+public class ReportServiceImpl implements ReportService {
+    private static final String SEPARATOR = ",";
+    private static final String HEADER = "fruit,quantity";
+
+    @Override
+    public String generateReport() {
+        StringBuilder stringBuilder = new StringBuilder(HEADER);
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/service/ReportServiceImpl.java:11]: do not use such naming(stringBuilder), how to guess what it stores?
```suggestion
        StringBuilder report = new StringBuilder(HEADER);
```

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
left a few comments

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
Left a few comments

**Review by Ivan95kos — APPROVED** [review]
Nice
